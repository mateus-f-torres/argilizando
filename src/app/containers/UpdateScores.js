import { connect } from 'react-redux';
import { resetScore, incrementScore, decrementScore, toggleDescription } from '../actions';
import { doneScore, showMenu } from '../actions';
import AbilityScores from '../components/AbilityScores.jsx';

const mapStateToProps = (state) => {
  return {
    total: state.score.total,
    str: state.score.str,
    dex: state.score.dex,
    con: state.score.con,
    int: state.score.int,
    wis: state.score.wis,
    cha: state.score.cha
  };
};

const mapDispatchToProps = (dispatch) => {
  return {
    scorePlus: (ability) => {
      dispatch(incrementScore(ability))
    },
    scoreMinus: (ability) => {
      dispatch(decrementScore(ability))
    },
    toggleText: (ability) => {
      dispatch(toggleDescription(ability))
    },
    resetBtn: () => {
      dispatch(resetScore())
    },
    lockScore: () => {
      dispatch(doneScore())
      dispatch(showMenu())
    }
  }
}

const UpdateScores = connect(
  mapStateToProps,
  mapDispatchToProps
)(AbilityScores);

export default UpdateScores;
