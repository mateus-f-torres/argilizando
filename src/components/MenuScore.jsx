//@flow
import * as React from 'react';
import {Link} from 'react-router-dom';
import CardScore from './CardScore.jsx';
import ABILITIES from '../data/scores.js';
import type {LongScore, ShortScore} from '../types/props.js';

import 'Styles/MenuScore.scss';

type Props = {
  total: number,
  str: LongScore,
  dex: LongScore,
  con: LongScore,
  int: LongScore,
  wis: LongScore,
  cha: LongScore,
  scorePlus: (string) => void,
  scoreMinus: (string) => void,
  toggleText: (string) => void,
  resetBtn: () => void,
  lockScore: (score: ShortScore) => void
};

class MenuScore extends React.Component<Props> {
  constructor(props: Props) {
    super(props);
    (this: any).handleClick = this.handleClick.bind(this);
    (this: any).resetBtn = this.resetBtn.bind(this);
    (this: any).lockScore = this.lockScore.bind(this);
  }

  handleClick(e: SyntheticEvent<HTMLButtonElement>) {
    e.preventDefault();
    let type = e.currentTarget.name;
    if (e.currentTarget.parentNode && e.currentTarget.parentNode.childNodes) {
      let h5 = e.currentTarget.parentNode.childNodes[1];
      // $FlowFixMe
      let ability = h5.attributes.name.value;
      switch (type) {
      case '+':
        this.props.scorePlus(ability);
        break;
      case '-':
        this.props.scoreMinus(ability);
        break;
      case '?':
        this.props.toggleText(ability);
        break;
      }
    }
  }

  resetBtn() {
    this.props.resetBtn();
  }

  lockScore() {
    let score = {
      str: [this.props.str[0], this.props.str[1]],
      dex: [this.props.dex[0], this.props.dex[1]],
      con: [this.props.con[0], this.props.con[1]],
      int: [this.props.int[0], this.props.int[1]],
      wis: [this.props.wis[0], this.props.wis[1]],
      cha: [this.props.cha[0], this.props.cha[1]],
    };
    this.props.lockScore(score);
  }

  render() {
    return (
      <section id="SCORE">
        <h2>Ability Scores</h2>
        <h3>{this.props.total} points left to spend</h3>
        <div className="horizontal-line" />
        <ul id="score-list">
          {
            // store unloads all six ABILITIES as props
            // e.g. props.str, props.dex, props.con
            // create a single score component for each
            // getting the correct score and modifier
            ABILITIES.map((name) => {
              let [score, mod, cost, show] = [...this.props[name[0]]];

              return (<CardScore
                key={name[0]}
                short={name[0]}
                long={name[1]}
                score={score}
                mod={mod}
                cost={cost}
                show={show}
                handleClick={this.handleClick}
              />);
            })
          }
        </ul>
        <div className="horizontal-line no-bottom" />
        <div className="lock-box">
          <button
            data-test="score-reset"
            onClick={this.resetBtn}>Reset</button>
          <Link to="/">
            <button
              data-test="score-lock"
              onClick={this.lockScore}>Lock</button>
          </Link>
        </div>
      </section>
    );
  }
}

export default MenuScore;
