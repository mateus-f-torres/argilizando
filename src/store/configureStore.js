import {createStore} from 'redux';
import rootReducer from '../reducers/index.js';

export default function configureStore() {
  return createStore(rootReducer);
}
