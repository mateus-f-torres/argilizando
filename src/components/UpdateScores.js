//@flow
import {connect} from 'react-redux';
import {withRouter} from 'react-router-dom';
import {
  incrementScore, decrementScore,
  toggleDescription, resetScore,
} from 'Store/score/actions';
import {lockScore} from 'Store/done/actions';
import MenuScore from 'Components/UpdateScores/MenuScore.js';
import type {State} from 'Types/state.js';

const mapStateToProps = (state: State) => {
  return {
    total: state.score.total,
    str: state.score.str,
    dex: state.score.dex,
    con: state.score.con,
    int: state.score.int,
    wis: state.score.wis,
    cha: state.score.cha,
  };
};

const mapDispatchToProps = (dispatch: *) => {
  return {
    scorePlus: (ability) => {
      dispatch(incrementScore(ability));
    },
    scoreMinus: (ability) => {
      dispatch(decrementScore(ability));
    },
    toggleText: (ability) => {
      dispatch(toggleDescription(ability));
    },
    resetBtn: () => {
      dispatch(resetScore());
    },
    lockScore: (score) => {
      dispatch(lockScore(score));
    },
  };
};

const UpdateScores = withRouter(connect(
  mapStateToProps,
  mapDispatchToProps
)(MenuScore));

export default UpdateScores;
