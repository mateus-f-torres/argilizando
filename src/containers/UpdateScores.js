import {connect} from 'react-redux';
import {withRouter} from 'react-router-dom';
import {
  incrementScore, decrementScore, toggleDescription,
  resetScore, lockScore,
} from '../actions';
import MenuScore from '../components/MenuScore.jsx';

const mapStateToProps = (state) => {
  return {
    total: state.score.total,
    str: state.score.str,
    dex: state.score.dex,
    con: state.score.con,
    int: state.score.int,
    wis: state.score.wis,
    cha: state.score.cha,
  };
};

const mapDispatchToProps = (dispatch) => {
  return {
    scorePlus: (ability) => {
      dispatch(incrementScore(ability));
    },
    scoreMinus: (ability) => {
      dispatch(decrementScore(ability));
    },
    toggleText: (ability) => {
      dispatch(toggleDescription(ability));
    },
    resetBtn: () => {
      dispatch(resetScore());
    },
    lockScore: (score) => {
      dispatch(lockScore(score));
    },
  };
};

const UpdateScores = withRouter(connect(
  mapStateToProps,
  mapDispatchToProps
)(MenuScore));

export default UpdateScores;
