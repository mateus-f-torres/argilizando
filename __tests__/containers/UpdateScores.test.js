import React from 'react';
import {Provider} from 'react-redux';
import {MemoryRouter} from 'react-router-dom';

import configureStore from 'redux-mock-store';
import {mount} from 'enzyme';

import UpdateScores from 'Containers/UpdateScores.js';
import MenuScore from 'Components/MenuScore.jsx';

const mockState = {
  score: {
    total: 45,
    str: [8, -1, 1, false],
    dex: [8, -1, 1, false],
    con: [8, -1, 1, false],
    int: [8, -1, 1, false],
    wis: [8, -1, 1, false],
    cha: [8, -1, 1, false],
  },
};

const mockStore = configureStore();
const store = mockStore(mockState);

describe('UpdateScores', () => {
  beforeEach(() => {
    store.clearActions();
  });

  const container = mount(
    <Provider store={store}>
      <MemoryRouter initialEntries={[{pathname: '/score', key: 'testKey'}]}>
        <UpdateScores />
      </MemoryRouter>
    </Provider>
  );

  describe('Unit Test', () => {
    test('render MenuScore component', () => {
      expect(container.containsMatchingElement(<MenuScore />)).toBe(true);
    });

    test('receive all props from store.score', () => {
      const component = container.find('MenuScore');

      expect(component.prop('total')).toBe(mockState.score.total);
      expect(component.prop('str')).toEqual(mockState.score.str);
      expect(component.prop('dex')).toEqual(mockState.score.dex);
      expect(component.prop('con')).toEqual(mockState.score.con);
      expect(component.prop('int')).toEqual(mockState.score.int);
      expect(component.prop('wis')).toEqual(mockState.score.wis);
      expect(component.prop('cha')).toEqual(mockState.score.cha);
    });

    test('dispatch incrementScore', () => {
      const component = container.find('MenuScore');

      const strPlusBtn = component.find('CardScore').at(0)
        .find('button').filter('[name="+"]');

      const dexPlusBtn = component.find('CardScore').at(1)
        .find('button').filter('[name="+"]');

      const expectedActions = [
        {
          'type': 'INCREMENT_ABILITY_SCORE',
          'payload': 'str',
        },
        {
          'type': 'INCREMENT_ABILITY_SCORE',
          'payload': 'dex',
        },
      ];

      strPlusBtn.simulate('click');
      dexPlusBtn.simulate('click');
      expect(store.getActions()).toEqual(expectedActions);
    });

    test('dispatch decrementScore', () => {
      const component = container.find('MenuScore');

      const conMinusBtn = component.find('CardScore').at(2)
        .find('button').filter('[name="-"]');

      const intMinusBtn = component.find('CardScore').at(3)
        .find('button').filter('[name="-"]');

      const expectedActions = [
        {
          'type': 'DECREMENT_ABILITY_SCORE',
          'payload': 'con',
        },
        {
          'type': 'DECREMENT_ABILITY_SCORE',
          'payload': 'int',
        },
      ];

      conMinusBtn.simulate('click');
      intMinusBtn.simulate('click');
      expect(store.getActions()).toEqual(expectedActions);
    });

    test('dispatch toggleDescription', () => {
      const component = container.find('MenuScore');

      const wisToggleBtn = component.find('CardScore').at(4)
        .find('button').filter('[name="?"]');

      const chaToggleBtn = component.find('CardScore').at(5)
        .find('button').filter('[name="?"]');

      const expectedActions = [
        {
          'type': 'TOGGLE_ABILITY_SCORE_DESCRIPTION',
          'payload': 'wis',
        },
        {
          'type': 'TOGGLE_ABILITY_SCORE_DESCRIPTION',
          'payload': 'cha',
        },
      ];

      wisToggleBtn.simulate('click');
      chaToggleBtn.simulate('click');
      expect(store.getActions()).toEqual(expectedActions);
    });

    test('dispatch resetScore', () => {
      const component = container.find('MenuScore');

      const resetBtn = component.find('[data-test="score-reset"]');

      const expectedActions = [
        {
          'type': 'RESET_ALL_ABILITY_SCORES',
        },
      ];

      resetBtn.simulate('click');
      expect(store.getActions()).toEqual(expectedActions);
    });

    test('dispatch lockScore', () => {
      const component = container.find('MenuScore');

      const lockBtn = component.find('[data-test="score-lock"]');

      const expectedActions = [
        {
          'type': 'LOCK_SCORE',
          'payload': {
            'str': [8, -1],
            'dex': [8, -1],
            'con': [8, -1],
            'int': [8, -1],
            'wis': [8, -1],
            'cha': [8, -1],
          },
        },
      ];

      lockBtn.simulate('click');
      expect(store.getActions()).toEqual(expectedActions);
    });
  });

  describe('Snapshot Test', () => {
    expect(container).toMatchSnapshot();
  });
});
