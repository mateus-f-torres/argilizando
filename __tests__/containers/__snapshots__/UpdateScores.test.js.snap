// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <MemoryRouter
    initialEntries={
      Array [
        Object {
          "key": "testKey",
          "pathname": "/score",
        },
      ]
    }
  >
    <Router
      history={
        Object {
          "action": "POP",
          "block": [Function],
          "canGo": [Function],
          "createHref": [Function],
          "entries": Array [
            Object {
              "hash": "",
              "key": "testKey",
              "pathname": "/score",
              "search": "",
            },
          ],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "index": 0,
          "length": 1,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "key": "testKey",
            "pathname": "/score",
            "search": "",
          },
          "push": [Function],
          "replace": [Function],
        }
      }
    >
      <withRouter(Connect(MenuScore))>
        <Route>
          <Connect(MenuScore)
            history={
              Object {
                "action": "POP",
                "block": [Function],
                "canGo": [Function],
                "createHref": [Function],
                "entries": Array [
                  Object {
                    "hash": "",
                    "key": "testKey",
                    "pathname": "/score",
                    "search": "",
                  },
                ],
                "go": [Function],
                "goBack": [Function],
                "goForward": [Function],
                "index": 0,
                "length": 1,
                "listen": [Function],
                "location": Object {
                  "hash": "",
                  "key": "testKey",
                  "pathname": "/score",
                  "search": "",
                },
                "push": [Function],
                "replace": [Function],
              }
            }
            location={
              Object {
                "hash": "",
                "key": "testKey",
                "pathname": "/score",
                "search": "",
              }
            }
            match={
              Object {
                "isExact": false,
                "params": Object {},
                "path": "/",
                "url": "/",
              }
            }
          >
            <MenuScore
              cha={
                Array [
                  8,
                  -1,
                  1,
                  false,
                ]
              }
              con={
                Array [
                  8,
                  -1,
                  1,
                  false,
                ]
              }
              dex={
                Array [
                  8,
                  -1,
                  1,
                  false,
                ]
              }
              history={
                Object {
                  "action": "POP",
                  "block": [Function],
                  "canGo": [Function],
                  "createHref": [Function],
                  "entries": Array [
                    Object {
                      "hash": "",
                      "key": "testKey",
                      "pathname": "/score",
                      "search": "",
                    },
                  ],
                  "go": [Function],
                  "goBack": [Function],
                  "goForward": [Function],
                  "index": 0,
                  "length": 1,
                  "listen": [Function],
                  "location": Object {
                    "hash": "",
                    "key": "testKey",
                    "pathname": "/score",
                    "search": "",
                  },
                  "push": [Function],
                  "replace": [Function],
                }
              }
              int={
                Array [
                  8,
                  -1,
                  1,
                  false,
                ]
              }
              location={
                Object {
                  "hash": "",
                  "key": "testKey",
                  "pathname": "/score",
                  "search": "",
                }
              }
              lockScore={[Function]}
              match={
                Object {
                  "isExact": false,
                  "params": Object {},
                  "path": "/",
                  "url": "/",
                }
              }
              resetBtn={[Function]}
              scoreMinus={[Function]}
              scorePlus={[Function]}
              str={
                Array [
                  8,
                  -1,
                  1,
                  false,
                ]
              }
              toggleText={[Function]}
              total={45}
              wis={
                Array [
                  8,
                  -1,
                  1,
                  false,
                ]
              }
            >
              <section
                id="SCORE"
              >
                <h2>
                  Ability Scores
                </h2>
                <h3>
                  45
                   points left to spend
                </h3>
                <div
                  className="horizontal-line"
                />
                <ul
                  id="score-list"
                >
                  <CardScore
                    cost={1}
                    handleClick={[Function]}
                    key="str"
                    long="Strength"
                    mod={-1}
                    score={8}
                    short="str"
                    show={false}
                  >
                    <li>
                      <h4>
                        Strength
                      </h4>
                      <h5
                        name="str"
                      >
                        <span
                          className="score"
                        >
                          8
                        </span>
                        <span
                          className="mod"
                        >
                          -1
                        </span>
                        Next lvl cost : 
                        <span
                          className="cost"
                        >
                          1
                        </span>
                      </h5>
                      <button
                        className="score-btn waves-effect waves-light"
                        data-test="score-plus"
                        name="+"
                        onClick={[Function]}
                      >
                        +
                      </button>
                      <button
                        className="score-btn waves-effect waves-light"
                        data-test="score-minus"
                        name="-"
                        onClick={[Function]}
                      >
                        -
                      </button>
                      <button
                        className="score-btn waves-effect waves-light"
                        data-test="score-show"
                        name="?"
                        onClick={[Function]}
                      >
                        ?
                      </button>
                    </li>
                  </CardScore>
                  <CardScore
                    cost={1}
                    handleClick={[Function]}
                    key="dex"
                    long="Dexterity"
                    mod={-1}
                    score={8}
                    short="dex"
                    show={false}
                  >
                    <li>
                      <h4>
                        Dexterity
                      </h4>
                      <h5
                        name="dex"
                      >
                        <span
                          className="score"
                        >
                          8
                        </span>
                        <span
                          className="mod"
                        >
                          -1
                        </span>
                        Next lvl cost : 
                        <span
                          className="cost"
                        >
                          1
                        </span>
                      </h5>
                      <button
                        className="score-btn waves-effect waves-light"
                        data-test="score-plus"
                        name="+"
                        onClick={[Function]}
                      >
                        +
                      </button>
                      <button
                        className="score-btn waves-effect waves-light"
                        data-test="score-minus"
                        name="-"
                        onClick={[Function]}
                      >
                        -
                      </button>
                      <button
                        className="score-btn waves-effect waves-light"
                        data-test="score-show"
                        name="?"
                        onClick={[Function]}
                      >
                        ?
                      </button>
                    </li>
                  </CardScore>
                  <CardScore
                    cost={1}
                    handleClick={[Function]}
                    key="con"
                    long="Constitution"
                    mod={-1}
                    score={8}
                    short="con"
                    show={false}
                  >
                    <li>
                      <h4>
                        Constitution
                      </h4>
                      <h5
                        name="con"
                      >
                        <span
                          className="score"
                        >
                          8
                        </span>
                        <span
                          className="mod"
                        >
                          -1
                        </span>
                        Next lvl cost : 
                        <span
                          className="cost"
                        >
                          1
                        </span>
                      </h5>
                      <button
                        className="score-btn waves-effect waves-light"
                        data-test="score-plus"
                        name="+"
                        onClick={[Function]}
                      >
                        +
                      </button>
                      <button
                        className="score-btn waves-effect waves-light"
                        data-test="score-minus"
                        name="-"
                        onClick={[Function]}
                      >
                        -
                      </button>
                      <button
                        className="score-btn waves-effect waves-light"
                        data-test="score-show"
                        name="?"
                        onClick={[Function]}
                      >
                        ?
                      </button>
                    </li>
                  </CardScore>
                  <CardScore
                    cost={1}
                    handleClick={[Function]}
                    key="int"
                    long="Intelligence"
                    mod={-1}
                    score={8}
                    short="int"
                    show={false}
                  >
                    <li>
                      <h4>
                        Intelligence
                      </h4>
                      <h5
                        name="int"
                      >
                        <span
                          className="score"
                        >
                          8
                        </span>
                        <span
                          className="mod"
                        >
                          -1
                        </span>
                        Next lvl cost : 
                        <span
                          className="cost"
                        >
                          1
                        </span>
                      </h5>
                      <button
                        className="score-btn waves-effect waves-light"
                        data-test="score-plus"
                        name="+"
                        onClick={[Function]}
                      >
                        +
                      </button>
                      <button
                        className="score-btn waves-effect waves-light"
                        data-test="score-minus"
                        name="-"
                        onClick={[Function]}
                      >
                        -
                      </button>
                      <button
                        className="score-btn waves-effect waves-light"
                        data-test="score-show"
                        name="?"
                        onClick={[Function]}
                      >
                        ?
                      </button>
                    </li>
                  </CardScore>
                  <CardScore
                    cost={1}
                    handleClick={[Function]}
                    key="wis"
                    long="Wisdom"
                    mod={-1}
                    score={8}
                    short="wis"
                    show={false}
                  >
                    <li>
                      <h4>
                        Wisdom
                      </h4>
                      <h5
                        name="wis"
                      >
                        <span
                          className="score"
                        >
                          8
                        </span>
                        <span
                          className="mod"
                        >
                          -1
                        </span>
                        Next lvl cost : 
                        <span
                          className="cost"
                        >
                          1
                        </span>
                      </h5>
                      <button
                        className="score-btn waves-effect waves-light"
                        data-test="score-plus"
                        name="+"
                        onClick={[Function]}
                      >
                        +
                      </button>
                      <button
                        className="score-btn waves-effect waves-light"
                        data-test="score-minus"
                        name="-"
                        onClick={[Function]}
                      >
                        -
                      </button>
                      <button
                        className="score-btn waves-effect waves-light"
                        data-test="score-show"
                        name="?"
                        onClick={[Function]}
                      >
                        ?
                      </button>
                    </li>
                  </CardScore>
                  <CardScore
                    cost={1}
                    handleClick={[Function]}
                    key="cha"
                    long="Charisma"
                    mod={-1}
                    score={8}
                    short="cha"
                    show={false}
                  >
                    <li>
                      <h4>
                        Charisma
                      </h4>
                      <h5
                        name="cha"
                      >
                        <span
                          className="score"
                        >
                          8
                        </span>
                        <span
                          className="mod"
                        >
                          -1
                        </span>
                        Next lvl cost : 
                        <span
                          className="cost"
                        >
                          1
                        </span>
                      </h5>
                      <button
                        className="score-btn waves-effect waves-light"
                        data-test="score-plus"
                        name="+"
                        onClick={[Function]}
                      >
                        +
                      </button>
                      <button
                        className="score-btn waves-effect waves-light"
                        data-test="score-minus"
                        name="-"
                        onClick={[Function]}
                      >
                        -
                      </button>
                      <button
                        className="score-btn waves-effect waves-light"
                        data-test="score-show"
                        name="?"
                        onClick={[Function]}
                      >
                        ?
                      </button>
                    </li>
                  </CardScore>
                </ul>
                <div
                  className="horizontal-line no-bottom"
                />
                <div
                  className="lock-box"
                >
                  <button
                    data-test="score-reset"
                    onClick={[Function]}
                  >
                    Reset
                  </button>
                  <Link
                    to="/"
                  >
                    <a
                      href="/"
                      onClick={[Function]}
                    >
                      <button
                        data-test="score-lock"
                        onClick={[Function]}
                      >
                        Lock
                      </button>
                    </a>
                  </Link>
                </div>
              </section>
            </MenuScore>
          </Connect(MenuScore)>
        </Route>
      </withRouter(Connect(MenuScore))>
    </Router>
  </MemoryRouter>
</Provider>
`;
